## Khan Academy Math Student Dashboard: Data Cleaning
We will be creating a student dashboard that displays each student's progress through Khan Academy Math. The goal is to showcase:
- How the student is performing overall and this month?
- What is their assignment completion rate?
- Which topics are students strongest and weakest in?
- The students' percentile in the class?
- Is the student growing?

We want to automate the dashboarding process for each class in Lyceum Village, grades K-8. We will be testing on grade K-1 students' Khan Academy Math first to create functions and a webpage to showcase their progress.

### Import the Data
The data imported is for grade K-1. The students' names are already adjusted to an alias name.

In [2]:
import numpy as np
import pandas as pd

In [16]:
df = pd.read_csv("../Resources/K_1-Math-Anonymous.csv")

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 266 entries, 0 to 265
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Assignment Name              266 non-null    object 
 1   Student Name                 266 non-null    object 
 2   Score At Due Date            26 non-null     float64
 3   Score Best Ever              30 non-null     float64
 4   Points Possible              30 non-null     float64
 5   Number Of Attempts           266 non-null    int64  
 6   Most Recent Completion Date  30 non-null     object 
 7   Start Date                   266 non-null    object 
 8   Due Date                     266 non-null    object 
 9   Assignment URL               266 non-null    object 
 10  Assignment Type              266 non-null    object 
 11  Due Date (no time)           266 non-null    object 
dtypes: float64(3), int64(1), object(8)
memory usage: 25.1+ KB


Our csv dataset will have 11 variables.

Types of data:
- 3 float: 
    - Score At Due Date
    - Score Best Ever
    - Points Possible
- 1 integer:
    - Number Of Attempts
- 7 string:
    - Assignment Name
    - Student Name
    - Most Recent Completion Date
    - Start Date
    - Due Date
    - Assignment URL
    - Assignment Type

The data type so far matches what they should be.

There are 266 rows total. There are missing data in the following columns:
- Score At Due Date: 240 missing
- Score Best Ever: 224 missing
- Points Possible: 224 missing
- Most Recent Completion Date: 224 missing

They are missing data because some students did not complete their assignments. We will be removing these data in actual student analysis.

### Student Anonymity
We will need to create a student ID column to give each student anonymity on their scores when it is being displayed to other parents.
Their ID's will be formatted as so: grade_studentinitial_number <br>

Example: G2_KL_1

#### Create Student ID
We will be automating this system by having it done automatically only once. The application will request the grade level from the user and store the student ID information into a PostgreSQL database.

In [5]:
df['Student Name'].unique()

array(['Ned Ethans', 'Anna Kite', 'April Luna', 'Andy Hong',
       'Abby Nguyen', 'Everett Chase', 'Kyle Anderson', 'Kevin Martelle',
       'Misa Bing', 'Olsen Le', 'Terry Long'], dtype=object)

In [6]:
# User input grade level
stu_grade = input("Enter the student grade level: ")

In [7]:
# Loop through the list of unique students and create their ID
stu_id_list = []
for name in df['Student Name']:
    stu_name = name.split(" ")
    stu_id = "G" + stu_grade + "_" + stu_name[0][0] + stu_name[1][0]
    stu_id_list.append(stu_id)
stu_id_list[:5]

['GK1_NE', 'GK1_AK', 'GK1_AL', 'GK1_AH', 'GK1_AN']

In [11]:
# Add in the numeric value to account for students with the same initials
stu_id_list2 = []
if len(df['Student Name']) == len(stu_id_list):
    for id in stu_id_list:
        # Add 1 at the end because there are no repeated initials
        stu_id_list2.append(id + "1")
else:
    # change the number depending on the student name
    # current list of students do not have repeated initials. Create a dummy list to test this function on
    stu_id_list2
stu_id_list2[:5]

['GK1_NE1', 'GK1_AK1', 'GK1_AL1', 'GK1_AH1', 'GK1_AN1']

In [17]:
# Add id column to the front of the dataframe
df.insert(loc=0, column='Student ID', value=stu_id_list2)
df.head()

Student ID          Assignment Name Student Name  Score At Due Date  \
0    GK1_NE1  Comparing numbers to 10   Ned Ethans                NaN   
1    GK1_AK1            Add within 10    Anna Kite                NaN   
2    GK1_AL1            Add within 10   April Luna                NaN   
3    GK1_AH1            Add within 10    Andy Hong                NaN   
4    GK1_AN1            Add within 10  Abby Nguyen                NaN   

   Score Best Ever  Points Possible  Number Of Attempts  \
0              NaN              NaN                   0   
1              NaN              NaN                   0   
2              NaN              NaN                   0   
3              NaN              NaN                   0   
4              NaN              NaN                   0   

  Most Recent Completion Date        Start Date          Due Date  \
0                         NaN  Sep 29th, 4:50PM  Oct 8th, 11:59PM   
1                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
2                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
3                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   
4                         NaN  Sep 29th, 4:48PM  Oct 8th, 11:59PM   

                                      Assignment URL Assignment Type  \
0  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
1  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
2  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
3  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
4  https://www.khanacademy.org/math/cc-kindergart...        Exercise   

  Due Date (no time)  
0          Oct 8th,   
1          Oct 8th,   
2          Oct 8th,   
3          Oct 8th,   
4          Oct 8th,

## Missing Data
NaN in the Score At Due Date, Score Best Ever, and Points Possible columns means that students did not complete their assignment. Thus, we will need to create a column that states the completion of their assignment.

Create a column with these values:
- Missing: Has 0 as the Number Of Attempts
- Complete: Has 1 or higher Number of Attempts and the Most Recent Completion Date is before or on the Due Date
- Late: Has 1 or more Number Of Attempts and the Most Recent Completion Date is past the Due Date

### Convert Date Columns to Date type
Convert the two columns from string object to date:
- Most Recent Completion Date
- Due Date

In [42]:
# Find the rows with no missing data
df_val = df[df['Most Recent Completion Date'].isna()==False]
df_val.head()

Student ID      Assignment Name    Student Name  Score At Due Date  \
96     GK1_KM1  Counting: Unit test  Kevin Martelle               13.0   
100    GK1_TL1  Counting: Unit test      Terry Long               13.0   
107    GK1_KM1         Add within 5  Kevin Martelle                7.0   
111    GK1_TL1         Add within 5      Terry Long                7.0   
118    GK1_KM1    Subtract within 5  Kevin Martelle                7.0   

     Score Best Ever  Points Possible  Number Of Attempts  \
96              13.0             13.0                   2   
100             13.0             13.0                   1   
107              7.0              7.0                   1   
111              7.0              7.0                   1   
118              7.0              7.0                   1   

    Most Recent Completion Date         Start Date           Due Date  \
96             Sep 20th, 4:55PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
100            Sep 20th, 4:42PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
107            Sep 20th, 5:01PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
111            Sep 20th, 4:50PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   
118            Sep 20th, 5:04PM  Sep 10th, 10:33AM  Sep 24th, 11:59PM   

                                        Assignment URL Assignment Type  \
96   https://www.khanacademy.org/math/early-math/cc...       Unit Test   
100  https://www.khanacademy.org/math/early-math/cc...       Unit Test   
107  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
111  https://www.khanacademy.org/math/cc-kindergart...        Exercise   
118  https://www.khanacademy.org/math/cc-kindergart...        Exercise   

    Due Date (no time)  
96          Sep 24th,   
100         Sep 24th,   
107         Sep 24th,   
111         Sep 24th,   
118         Sep 24th,

In [47]:
import datetime
test_due_date = df_val['Due Date'][96]
test_comp_date = df_val['Most Recent Completion Date'][96]

In [48]:
test_comp_date

'Sep 20th, 4:55PM'

In [49]:
test_due_date

'Sep 24th, 11:59PM'

In [52]:
import re

In [57]:
test_split1 = re.split(", | ", test_due_date)
test_split1

['Sep', '24th', '11:59PM']

In [58]:
re.findall('\d+', test_split1[1])[0]

'24'

In [ ]:
# Function to convert the date from string type to date type
import datetime
import re
def convertDate(date_str):
    split_date = re.split(", | ", date_str) # Splits the date into Month, Day, Time
    day_num = int(re.findall('\d+', split_date[1])[0]) # Removes letters from the day. Example: 8th becomes 8
    

In [28]:
progress_list = []
index_attempts = df.columns.get_loc('Number Of Attempts')
index_duedate = df.columns.get_loc('Due Date')
index_compdate = df.columns.get_loc('Most Recent Completion Date')
for i in df:
    if i[index_attempts] == 0:
        progress_list.append('Missing')
    else:
        if i[index_duedate] > i[index_compdate]:
            progress_list.append('Complete')
        else:
            progress_list.append('Late')

IndexError: string index out of range

In [30]:
import math
df['Number'].apply(lambda x: math.isnan(x)) == False

TypeError: must be real number, not str

In [29]:
df.loc[df['Most Recent Completion Date'].isnan() = False]

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: Most Recent Completion Date, dtype: object

In [35]:
df.dtypes

Student ID                      object
Assignment Name                 object
Student Name                    object
Score At Due Date              float64
Score Best Ever                float64
Points Possible                float64
Number Of Attempts               int64
Most Recent Completion Date     object
Start Date                      object
Due Date                        object
Assignment URL                  object
Assignment Type                 object
Due Date (no time)              object
dtype: object